lib/ansible/modules/system/vdo.py

```diff
@@ -0,0 +1,762 @@
+
+    if (('blockmapcachesize' in params) and
+            (params['blockmapcachesize'] is not None)):
+        options.append("--blockMapCacheSize=" + params['blockmapcachesize'])
+
+    if ('readcache' in params) and (params['readcache'] == 'enabled'):
+        options.append("--readCache=enabled")
+
+    if ('readcachesize' in params) and (params['readcachesize'] is not None):
+        options.append("--readCacheSize=" + params['readcachesize'])
+
+    if ('writepolicy' in params) and (params['writepolicy'] == 'async'):
+        options.append("--writePolicy=async")
+
+    if ('slabsize' in params) and (params['slabsize'] is not None):
+        options.append("--vdoSlabSize=" + params['slabsize'])
+
+    if ('emulate512' in params) and (params['emulate512'] == 'enabled'):
+        options.append("--emulate512=enabled")
+
+    if ('indexmem' in params) and (params['indexmem'] is not None):
+        options.append("--indexMem=" + params['indexmem'])
+
+    if ('indexmode' in params) and (params['indexmode'] == 'sparse'):
+        options.append("--sparseIndex=enabled")
+
+    # Entering an invalid thread config results in a cryptic
+    # 'Could not set up device mapper for %s' error from the 'vdo'
+    # command execution.  The dmsetup module on the system will
+    # output a more helpful message, but one would have to log
+    # onto that system to read the error.  For now, heed the thread
+    # limit warnings in the DOCUMENTATION section above.
+    if ('ackthreads' in params) and (params['ackthreads'] is not None):
+        options.append("--vdoAckThreads=" + params['ackthreads'])
+
+    if ('biothreads' in params) and (params['biothreads'] is not None):
+        options.append("--vdoBioThreads=" + params['biothreads'])
+
+    if ('cputhreads' in params) and (params['cputhreads'] is not None):
+        options.append("--vdoCpuThreads=" + params['cputhreads'])
+
+    if ('logicalthreads' in params) and (params['logicalthreads'] is not None):
+        options.append("--vdoLogicalThreads=" + params['logicalthreads'])
+
+    if (('physicalthreads' in params) and
+            (params['physicalthreads'] is not None)):
+        options.append("--vdoPhysicalThreads=" + params['physicalthreads'])
+
+    vdocmdoptions = ' '.join(options)
+    return vdocmdoptions
+
+
+def run_module():
+
+    # Debugging note:
+    # Don't use "print" in ansible modules.  To print debug messages,
+    # add it to the "result" dictionary, the contents of which is printed
+    # when executing "ansible-playbook foo.yml -vvvv".
+
+    # Define the available arguments/parameters that a user can pass to
+    # the module.
+    # Defaults for VDO parameters are None, in order to facilitate
+    # the detection of parameters passed from the playbook.
+    # Creation param defaults are determined by the creation section.
+
+    module_args = dict(
+        name=dict(type='str', required=True),
+        state=dict(choices=['absent', 'present'],
```

Comment:
There's an issue with changing a number of these arguments into booleans: there is no way to specify a default of "None" (i.e.: the user did not specify the playbook argument, therefore the parameter should not be changed from its previous state to some arbitrary default that the user may not want, and may not be able to track without active logging.) (Note that it is possible to specify "type='bool', default=None", but it is functionally equivalent to "type='bool', default=False".  This might be worth documenting for authors of future Ansible modules.)

The module arguments "activated", "running", "deduplication", "compression", and "readcache" shall remain a type of "choice", with a default of "None", to preserve consistent behavior for parameters of VDO volumes that can be modified.  The playbook intent should be respected here: if an optional "modify" parameter is not specified, do not change it away from its previous state.

(I'm saying this after having experimented, and wondering why a VDO volume was deactivated on creation, or mysteriously activated when I did not say it should be activated.  Please consider the systems implications behind a storage volume that the administrator wants disabled, but is suddenly enabled, or vice versa.)